### DISCUSSION TITLE
THẢO LUẬN BÁO CÁO BÀI 5 VÀ 6

### DISCUSSION THREAD

**Discussion by trthann** [trthann/L-p-10A06---Nh-m-02#9] [General]
Có bạn nào có thắc mắc gì không nhỉ ?

**Comment by trthann** [comment]
Câu 1 bài 5 :

Các loại ngôn ngữ lập trình chính: Python, Ruby, C, Java, Pascal,…. (Thầy Hiền dạy)
 	Python: nghiên về xử lí trí tuệ nhân tạo:
	Ưu điểm:
-	Là một ngôn ngữ có hình thức sáng sủa, cấu trúc roc ràng, cú pháp ngắn gọn
-	Có trên tất cả các nền tảng hệ điều hành từ UNIX, MS – DOS, Mac OS, Windows và Linix và các OS khác thuộc họ Unix.
-	Tương thích mạnh mẽ với Unix, hardware, thirt-party software với số lượng thư viện khổng lồ (400 triệu người sử dụng)
-	Python với tốc độ xử lý cực nhanh, python có thể tạo ra những chương trình từ những script siêu nhỏ tới những phần mềm cực lớn như Biender 3D.
	Nhược điểm:
-	Python không có các thuộc tính như :protected,private hay public, không có vòng lặp do…while và switch….case.
-	Mặc dù tốc độ xử lý của Python nhanh hơn PHP nhưng không bằng JAVA và C++.
(http://itplus-academy.edu.vn/Uu-va-nhuoc-diem-cua-Python-trong-lap-trinh.html)

 	Java
	Ưu điểm:
-	Con đường học tập: Con đường học tập đối với ngôn ngữ lập trình Java là cực kỳ ngắn. 
-	Lập trình hướng đối tượng: Ở đây có một lợi thế thú vị khác khi sử dụng Java. Đó là Java cho phép bạn tạo các chương trình mô-đun và mã có thể tái sử dụng để giữ cho hệ thống có thể mở rộng cũng như linh hoạt.
-	API phong phú: Java cung cấp các API cho các hoạt động khác nhau như kết nối cơ sở dữ liệu, kết nối mạng, I / O, phân tích cú pháp XML, các tiện ích và hơn thế nữa.
-	Công cụ mã nguồn mở mạnh mẽ phát triển nhanh chóng: Một số lượng lớn các công cụ phát triển mã nguồn mở sử dụng Java đã làm cho Java trở thành một lựa chọn mạnh mẽ hơn cho các nhà phát triển, ví dụ như Eclipse và Netbeans. 
-	Thư viện mã nguồn mởSố lượng lớn các thư viện nguồn mở và trưởng thành với sự hỗ trợ công nghiệp cũng đảm bảo rằng Java được sử dụng ở mọi nơi. Một số thư viện này đã được đóng góp bởi nhiều tổ chức phổ biến như Apache, Google, Yahoo, Facebook, LinkedIn,… Điều đó làm cho Java trở nên phổ biến hơn.
-	Miễn phí 
-	Hỗ trợ cộng đồng: Cuối cùng, có một cộng đồng hỗ trợ Java lớn mà ngôn ngữ này tập hợp được. Cọng đồng này sẽ giúp đồng nhành cùng các Java developer để học về nghệ thuật lập trình cũng như để làm việc hiệu quả hơn.

	Nhược điểm:
- Trình biên dịch Java chưa được tối ưu hóa tốt so với C ++.
- Không có sự tách biệt đặc điểm kỹ thuật khi triển khai.
- Quản lý bộ nhớ, với Java, là khá tốn kém.
- Việc thiếu các template có thể hạn chế khả năng của Java để tạo ra các cấu trúc dữ liệu chất lượng cao.
- Người ta có thể tìm thấy một số lỗi trong trình duyệt và các chương trình ví dụ.

(https://niithanoi.vn/chi-tiet-tin/1729/7-diem-manh-5-diem-yeu-cua-ngon-ngu-lap-trinh-java.html#.YYfL-mBByUl)


 	Ruby:
	Ưu điểm:
-	Công cụ: Rails cung cấp cho ta công cụ tuyệt vời giúp chúng ta triển khải được nhiều tính năng hơn mà lại tốn ít thời gian hơn. Nó cung cấp cho ta một cấu trúc chuẩn cho ứng dụng web.
-	Thư viện: Rails cung cấp cho ta gem, thứ có thẻ thay thế hầu hết cho mọi thứ mà bạn có thể nghĩ ra được. Hơn nữa, tất cả gem đều có thể sử dụng một cách hoàn toàn miền phí và có thể dễ dàng tra cứu tại https://rubygems.org/.
-	Chất lượng code: Nhìn chung, mình nhận thấy rằng chất lượng của các phần mềm hỗ trợ Ruby code là cao hơn so với một số ngôn ngữ khác.
-	Cộng đồng: Cộng đồng Ruby rất lớn. Mình tin rằng điều này giúp cải thiện những sản phẩm của Ruby rất nhiều và đây cũng là một lý do mà thư viện của Ruby lại tuyệt vời như vậy. Ruby cũng là một trong số những ngôn ngữ lập trình phổ biến nhất trên Github.
-	Hiệu suất: RoR là một ngôn ngữ gọn gàng, khi mà sử dụng kết hợp cùng các thư viện hỗ trợ, RoR cho phép bạn phát triển một cách khá là nhanh chóng.
-	Tương lai: RoR có thể được xem là ngôn ngữ được lựa chọn trong các trang học lập trình trực tuyến phổ biến như Makers Academy, CodeCademy . Điều này có nghĩa là số lượng lập trình viên tài giỏi tham gia vào cộng đồng Ruby sẽ tăng lên trong các năm tới.
	Nhược điểm:
-	Thời gian chạy: Phần lớn những cuộc tranh luận về Rail đều có những ý kiến răng Rails chậm. Điều này hoàn toàn đúng, so với NodeJS hay GoLang, Rails có tốc độ chạy khá chậm. Mặc dù trong thực thế, hiệu suất của ứng dụng Rails rất khó làm chậm lại việc kinh doanh, Vấn đề thường nằm ở việc thiết kế server, thậm chí là cả cơ sở dữ liệu.
-	Tốc độ boot: Vấn đề chính gây khó chịu khi làm việc với Rails là tốc độ boot của Rails framework. Phụ thuộc vào số lượng gem và file, chúng ta có thể mất một khoảng thời gian kha khá để khởi động, điều này có thể ảnh hưởng trực tiếp tới hiệu suất làm việc. Với việc Spring ra đời, tốc độ boot đã được cải thiện nhưng có vẻ như nó vẫn còn có thể nhanh hơn nữa.
(https://viblo.asia/p/uu-nhuoc-diem-cua-ruby-on-rails-eXaRlwDoGmx)

**Comment by trthann** [comment]
Câu 2 bài 5:
Trong loại Ngôn ngữ bậc cao, 3 ngôn được yêu thích nhất là: Python, Java, Ruby:
Vì: 
-	Các ngôn ngữ trên ứng dụng vào trí tuệ nhân tạo góp ích cho cuộc sống hiện nay.
-	Được thế giới ưu chuộng và dung rất nhiều.
-	Có thời gian phát triển dài, có tiếng tăm trong ngành cộng nghệ
-	Đóng góp bởi nhiều tổ chức phổ biến như Apache, Google, Yahoo, Facebook, LinkedIn,…
-	Dễ dàng đọc và dễ sử dụng

**Comment by PhuongNgannLe** [comment]
Câu 3 Bài 6:
Việc giải bài toán trên máy tính không giống như việc giải quyết vấn đề ngoài đời thực . Tại vì máy tính chỉ có thể xử lý các bài toán đã được lập trình sẵn các bước , có sẵn thuật toán và sử dụng ngôn ngữ riêng của máy tính. 
Bài 1 : Cho phương trình ax^2 + bx + c = 0 . Tìm nghiệm của phương trình 
Bài 2 : Cho số nguyên dương A. Hãy tính A là số chẵn hay số lẻ
Bài 3 : Cho 2 số nguyên dương A và B . Hãy tính ước chung lớn nhất của 2 số A , B đó

**Comment by PhuongNgannLe** [comment]
Câu 1 Bài 5:
Ngôn ngữ lập trình chính gồm có 3 loại: Ngôn ngữ máy, hợp ngữ và ngôn ngữ bậc cao
a. Ngôn ngữ máy
- Ưu điểm: 
+ Là ngôn ngữ duy nhất máy tính có thể trực tiếp hiểu và thực hiện, cho phép khai thác triệt để và tối ­ưu khả năng của máy. 
- Nhược điểm: 
+ Ngôn ngữ phức tạp, phụ thuộc nhiều vào phần cứng, ch­ương trình viết mất nhiều công sức, cồng kềnh và khó hiệu chỉnh.
+ Không thích hợp với số đông ng­ười lập trình.
b. Hợp ngữ
- Ưu điểm: 
+ Hợp ngữ cho phép người lập trình sử dụng một số từ (thường là các từ tiếng Anh viết tắt) để thể hiện các lệnh cần thực hiện.
+ Cho phép khai thác triệt để tính năng phần cứng.
- Nhược điểm: 
+ Thuận lợi cho các nhà lập trình chuyên nghiệp nhưng chưa thích hợp với số đông người lập trình.
+ Để máy tính hiểu cần có chương trình hợp dịch để chuyển hợp ngữ sang ngôn ngữ máy.
c. Ngôn ngữ bậc cao
- Ưu điểm:
+ Dễ hiểu, dễ chỉnh sửa, tính độc lập cao.
+ Để máy tính hiểu được, cần phải có chương trình dịch để chuyển từ ngôn ngữ bậc cao sang ngôn ngữ máy.
+ Một số ngôn ngữ lập trình bậc cao: Pascal, C, Java, …
         https://hoc24.vn/ly-thuyet/bai-5-ngon-ngu-lap-trinh.3225
